### PR TITLE
Fixes #845: gru init: prefer named github_hosts alias (name:owner/repo) over legacy host/owner/repo form

### LINKED ISSUE #845: gru init: prefer named github_hosts alias (name:owner/repo) over legacy host/owner/repo form
## Problem

`gru init` writes the new repo entry to `daemon.repos` using the legacy `host/owner/repo` form for any non-`github.com` host. This ignores named `[github_hosts.*]` config entries, so users with a named host (e.g. `netflix`) end up with `github.netflix.net/owner/repo` instead of `netflix:owner/repo` in their config, even though they've already configured the friendlier alias.

## Proposed fix

In `handle_init` (`src/commands/init.rs`), before falling back to `host/owner/repo`, look up `config.github_hosts` for an entry whose `host` matches (case-insensitive). If found, use `name:owner/repo`.

Resolution order:
1. `github.com` → `owner/repo`
2. Host matches an existing named `[github_hosts.*]` entry → `name:owner/repo`
3. Host contains a `.` → legacy `host/owner/repo`

## Example

Config already has:
```toml
[github_hosts.netflix]
host = "github.netflix.net"
```

Running `gru init github.netflix.net/foo/bar` should add `"netflix:foo/bar"` to `daemon.repos`, not `"github.netflix.net/foo/bar"`.

### PR DESCRIPTION
**Pull request by fotoetienne**
## Summary
- `gru init` now prefers a named `[github_hosts.*]` alias over the legacy `host/owner/repo` form when writing `daemon.repos`. With `[github_hosts.netflix] host = "github.netflix.net"` already configured, `gru init github.netflix.net/foo/bar` now adds `netflix:foo/bar` instead of `github.netflix.net/foo/bar`.
- Extracted the repo-entry formatting into a pure `build_repo_entry` helper in `src/commands/init.rs` and unit-tested each branch of the resolution order.
- Logs a warning (instead of silently ignoring) when config parsing fails during alias lookup, so a broken config can't quietly cause a legacy fallback.

Resolution order:
1. `github.com` → `owner/repo`
2. Host matches a named `[github_hosts.*]` entry (case-insensitive) → `name:owner/repo`
3. Host contains a dot → legacy `host/owner/repo`
4. No dot and no alias → warn and skip

Fixes #845.

## Test plan
- `cargo test commands::init` — 5 new unit tests cover github.com, named alias match, case-insensitive alias match, legacy fallback, and no-dot `None`.
- `just check` — fmt, clippy, all 1243 tests pass, release build succeeds.

## Notes
- `validate_github_hosts` already enforces that the `host` field is unique across `[github_hosts.*]` entries, so at most one alias can match a given host.
- If `LabConfig::load_partial` fails (malformed config), the code now logs a warning and falls back to the legacy form; previously it would silently fall back.

<sub>🤖 M1gx</sub>

### CODE REVIEW

**Review by fotoetienne — COMMENTED** [review]
## Review

Clean, well-scoped fix. The resolution order matches the issue requirement and the refactor into `build_repo_entry` is a nice testability win.

### Strengths
- Pure helper with dependency injection (`HashMap<String, GhHostConfig>`) makes the logic trivially testable — all 5 branches covered.
- Case-insensitive alias matching is consistent with how `host` is treated elsewhere (`eq_ignore_ascii_case` for `github.com`).
- Warning on `load_partial` failure (instead of silent fallback) is the right call — a malformed config shouldn't quietly downgrade the user to a legacy form they explicitly tried to avoid.
- Good reliance on `validate_github_hosts`'s uniqueness guarantee, noted in a comment.

### Minor observations (non-blocking)

1. **`String::new()` as sentinel for skip** — `build_repo_entry` returns `Option<String>`, but the `unwrap_or_else` closure returns `""` which is then checked via `!repo_entry.is_empty()`. Slightly awkward; a `match` or `if let Some(entry) = ...` would express intent more directly. Not worth changing.

2. **Warning message specificity** — the `unwrap_or_else` warning says "host … has no dot", which is accurate today because that's the only `None` branch. If a future branch returns `None` for another reason, the message could mislead. Fine for now given there's only one `None` path.

3. **Config reloaded between steps 3 and 4** — `write_default_config` may have just created the config, and then `load_partial` reads it. If there's any edge case where step 3 succeeds but step 4's parse fails, the user sees a confusing warning. Unlikely in practice.

### Test coverage
The 5 new unit tests cover github.com, named-alias match, case-insensitive match, legacy fallback, and no-dot `None`. Consider also a test where a `[github_hosts.*]` entry exists but its `host` doesn't match (should fall through to legacy) — though the legacy-fallback test with an empty map arguably covers the "no alias matches" logic already.

Overall: LGTM.

<sub>🤖 M1gz</sub>

**Comment by fotoetienne** [comment]
M1gz, thanks for the review! Addressed the two actionable items:

- **`String::new()` sentinel** — replaced with a `match` on `build_repo_entry`'s `Option<String>` result so the `None` branch warns directly and the `Some(entry)` branch feeds straight into `add_repo_to_config`. No more empty-string sentinel round-trip.
- **Non-matching alias test** — added `test_build_repo_entry_non_matching_named_host_falls_through_to_legacy`: `[github_hosts.netflix] host = "github.netflix.net"` present, but resolving `ghe.other.com` falls through to the legacy `ghe.other.com/foo/bar` form.

Skipped the warning-message nit and the reload-between-steps-3-and-4 observation since they're both conditional on hypothetical future changes. `just check` passes (1244 tests).

<sub>🤖 M1gx</sub>

**Comment by fotoetienne** [comment]
M1gz, thanks for the thorough review.

Addressed two of the non-blocking observations in 8a91e1c:

1. **`String::new()` sentinel** — replaced with a `match` on `Option<String>`, so the intent is expressed directly instead of via an empty-string check.
2. **Fall-through test** — added `test_build_repo_entry_falls_through_when_no_alias_matches` which configures a `netflix → github.netflix.net` alias and verifies that initializing `ghe.other.com/foo/bar` falls through to the legacy `ghe.other.com/foo/bar` form.

Left #2 (warning message specificity) and #3 (config reloaded between steps) as-is per your "fine for now" call — happy to revisit if a future branch adds another `None` path.

`just check` passes (1245 tests, 10 skipped).

<sub>🤖 M1gx</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the review. Addressed all three inline comments in b6871d8 — details on each inline thread.

High-level summary:
- **Case-sensitivity mismatch**: made `validate_github_hosts` compare hosts case-insensitively (DNS hostnames are case-insensitive), so the uniqueness guarantee now matches the case-insensitive lookup in `build_repo_entry`. Added a regression test.
- **Warning message**: reworded the `load_partial` failure warning to describe both branches (github.com / legacy-or-skip) rather than implying legacy fallback is always used.
- **CLI invocation in examples**: confirmed on the thread — `parse_repo_source` only accepts `owner/repo` or `.`, so the correct invocation is `gru init foo/bar --host github.netflix.net` (or `gru init .` from a checkout). The behavioral change is the same either way.

`just check` passes (1246 tests, 10 skipped).

<sub>🤖 M1gx</sub>

File: src/commands/init.rs [hunk 1]
```diff
@@ -82,6 +83,36 @@ fn validate_host(host: &str) -> Result<()> {
...
+) -> Option<String> {
+    if host.eq_ignore_ascii_case("github.com") {
+        return Some(format!("{}/{}", owner, repo));
+    }
+    // `validate_github_hosts` enforces that `host` is unique across entries,
+    // so at most one match is possible here.
+    if let Some((name, _)) = github_hosts
+        .iter()
+        .find(|(_, gh)| gh.host.eq_ignore_ascii_case(host))
+    {
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs]: Copilot, good catch — the mismatch was real. Rather than downgrading the lookup to case-sensitive, I made the validator case-insensitive (`to_ascii_lowercase` keys) since DNS hostnames are case-insensitive. `validate_github_hosts` in `src/config.rs` now rejects duplicates regardless of host casing, and the comment in `build_repo_entry` was updated to note the case-insensitive uniqueness guarantee. Added a regression test `test_validate_github_host_duplicate_host_case_insensitive` covering `ghe.example.com` vs `GHE.Example.COM`.

<sub>🤖 M1gx</sub>

**Comment by fotoetienne** [comment]
copilot-pull-request-reviewer[bot], thanks for the review summary. Addressed all three inline comments in b6871d8 — details in the per-thread replies below.

<sub>🤖 M1gx</sub>

File: src/commands/init.rs [hunk 2]
```diff
@@ -204,27 +235,19 @@ pub(crate) async fn handle_init(repo_arg: String, host_override: Option<String>)
...
-            repo,
-            host,
-        );
-        String::new()
+    let github_hosts = match LabConfig::load_partial(&config_path) {
+        Ok(cfg) => cfg.github_hosts,
+        Err(e) => {
+            log::warn!(
+                "  ⚠️  Could not load config for [github_hosts.*] lookup, \
+                 falling back to legacy host/owner/repo form: {:#}",
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs]: Copilot, fair point — the old wording implied the legacy form is always the fallback, which isn't true for `github.com` or dotless hosts. Reworded the message to describe all three outcomes: `owner/repo` for github.com, legacy `host/owner/repo` for other hosts with a dot, or skip for dotless hosts.

<sub>🤖 M1gx</sub>

File: src/commands/init.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs]: Copilot, good catch — taking the first option (enforce uniqueness case-insensitively in the validator), since DNS hostnames are case-insensitive and the rest of the codebase already treats them that way (e.g. `github.com` comparisons).

Changes in b6871d8:

- `validate_github_hosts` (`src/config.rs`) now keys both passes on the lowercased host, so two entries differing only by case (e.g. `GHE.Example.com` and `ghe.example.com`) are rejected as duplicates.
- Updated the comment in `build_repo_entry` to note the uniqueness guarantee is now explicitly case-insensitive, matching the lookup.
- Added `test_validate_github_host_duplicate_host_case_insensitive` to pin the new behavior.

1246 tests pass.

<sub>🤖 M1gx</sub>

File: src/commands/init.rs [hunk 3]
```diff
@@ -82,6 +83,36 @@ fn validate_host(host: &str) -> Result<()> {
...
+/// 4. Otherwise → `None` (caller should warn and skip)
+fn build_repo_entry(
+    host: &str,
+    owner: &str,
+    repo: &str,
+    github_hosts: &HashMap<String, GhHostConfig>,
+) -> Option<String> {
+    if host.eq_ignore_ascii_case("github.com") {
+        return Some(format!("{}/{}", owner, repo));
+    }
```
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs:101]: Copilot, you're right — `parse_repo_source` only accepts `owner/repo` or `.`, so `gru init github.netflix.net/foo/bar` would be rejected as ambiguous. The original issue example uses the same shorthand, but the actual CLI invocations that exercise this fix are:

- `gru init foo/bar --host github.netflix.net` (explicit host flag), or
- `gru init .` from an existing checkout whose remote is on `github.netflix.net`.

The behavioral change (prefer `netflix:foo/bar` over `github.netflix.net/foo/bar` in `daemon.repos`) is the same for both invocations. Not expanding `parse_repo_source` to accept `host/owner/repo` input in this PR — that's a separate CLI-surface change. Left the `build_repo_entry` code as-is and will correct the PR/issue examples in the merge-time description.

<sub>🤖 M1gx</sub>

File: src/commands/init.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs]: Copilot, agreed — the old wording implied a single outcome even though three are possible. Reworded in b6871d8 to drop the "falling back to legacy form" phrasing and instead enumerate what will happen: `owner/repo` for github.com, legacy `host/owner/repo` for other hosts with a dot, or skipped otherwise.

New text:

> Could not load config for [github_hosts.*] lookup; proceeding without alias resolution (repo entry will use owner/repo for github.com, legacy host/owner/repo for other hosts with a dot, or be skipped): {e}

<sub>🤖 M1gx</sub>

File: src/commands/init.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fotoetienne) [reply src/commands/init.rs:101]: Copilot, correct — this is a doc issue, not a code one. The issue body carried the `gru init github.netflix.net/foo/bar` wording and I propagated it into my PR description, but `parse_repo_source` only accepts `owner/repo` or `.`. The actual supported invocations that exercise this code path are:

- `gru init foo/bar --host github.netflix.net`
- `cd <repo-checkout> && gru init .` (host is detected from the git remote)

No CLI change is intended in this PR. Updated the minion-side PR description accordingly; leaving `parse_repo_source` unchanged since broadening it to accept `host/owner/repo` is out of scope.

<sub>🤖 M1gx</sub>

**Comment by fotoetienne** [comment]
---
type: monitoring-paused
---

⏸️ This PR's automated agent has paused. There are 6 reviews that haven't been addressed yet. Resume automated responses with:
`gru resume M1gx`

<sub>🤖 M1gx</sub>

**Comment by fotoetienne** [comment]
---
type: monitoring-paused
---

⏸️ This PR's automated agent has paused. There are 6 reviews that haven't been addressed yet. Resume automated responses with:
`gru resume M1gx`

<sub>🤖 M1gx</sub>
